astropy/units/decorators.py

```diff
@@ -100,62 +167,55 @@ def wrapper(*func_args, **func_kwargs):
                 if param.name not in bound_args.arguments and param.default is not param.empty:
                     bound_args.arguments[param.name] = param.default
 
                 # Get the value of this parameter (argument to new function)
                 arg = bound_args.arguments[param.name]
 
-                # Get target unit, either from decorator kwargs or annotations
+                # Get target unit or physical type, either from decorator kwargs
+                #   or annotations
                 if param.name in self.decorator_kwargs:
-                    target_unit = self.decorator_kwargs[param.name]
+                    targets = self.decorator_kwargs[param.name]
                 else:
-                    target_unit = param.annotation
+                    targets = param.annotation
 
-                # If the target unit is empty, then no unit was specified so we
-                # move past it
-                if target_unit is not funcsigs.Parameter.empty:
+                # If the targets is empty, then no target units or physical
+                #   types were specified so we can continue to the next arg
+                if targets is funcsigs.Parameter.empty:
+                    continue
 
-                    if isinstance(target_unit, str):
+                # If the argument value is None, and the default value is None,
+                #   pass through the None even if there is a target unit
+                if arg is None and param.default is None:
+                    continue
 
-                        try: # unit passed in as a string
-                            target_unit = Unit(target_unit)
-                            str_target_unit = target_unit.to_string()
-                        except ValueError:
-                            # user specified a physical type instead of a unit
-                            try:
-                                physical_type_id = _unit_physical_mapping[target_unit]
-                                str_target_unit = target_unit
-                            except KeyError:
-                                raise ValueError("Invalid unit of physical type '{0}'."
-                                                 .format(target_unit))
+                # Here, we check whether multiple target unit/physical type's
+                #   were specified in the decorator/annotation, or whether a
+                #   single string (unit or physical type) or a Unit object was
+                #   specified
+                if isinstance(targets, str) or not isiterable(targets):
+                    valid_targets = [targets]
 
-                            target_unit = Unit._from_physical_type_id(physical_type_id)
+                # Check for None in the supplied list of allowed units and, if
+                #   present and the passed value is also None, ignore.
+                elif None in targets:
+                    valid_targets = [t for t in targets] # make a shallow copy
```

Comment:
Good point about the copy -- otherwise we change the targets. But maybe just
```
if arg is None:
    continue
else:
    targets = [t for t in targets if t is not None
```